### PR TITLE
contrib: add QAT private key provider.

### LINKED ISSUE #21531: Add Intel® QuickAssist Technology (Intel® QAT) support as contrib extensions
*Title*:

Add Intel® QuickAssist Technology (Intel® QAT) support as contrib extensions

*Description*:

Next generation Intel® QAT support with Intel® Xeon® Scalable processors will feature an Intel® QAT cryptography and compression acceleration engine. In order to use it, two new Envoy contrib extensions are proposed:

1.	QATzip compression extension. This extension uses QATzip library (https://github.com/intel/QATzip) to implement accelerated HTTP compression. This can have a valid use case in scenarios such as service meshes, where the traffic within the mesh may be compressed without consuming CPU cycles.
2.	QAT private key provider. This extension will use qatlib library (https://github.com/intel/qatlib) to accelerate RSA operations in handshakes. The extension will look a bit like the existing cryptomb private key provider. The use case is to move the expensive cryptographic operations away from the CPU to the accelerator device, leaving CPU cycles for other use.

Two new library dependencies would be added: QATzip library for compression and qatlib for accessing the hardware. Support for Intel® QAT is already present in the mainline Linux kernel and in Kubernetes device plugins (to expose the device files to containers). There are previous generations of Intel QAT® hardware devices, but they are not supported by these extensions.

*Relevant links*:

https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/cmd/qat_plugin/README.md
https://www.intel.com/content/www/us/en/architecture-and-technology/intel-quick-assist-technology-overview.html
https://www.intel.com/content/www/us/en/developer/topic-technology/open/quick-assist-technology/overview.html

### PR DESCRIPTION
**Pull request by ipuustin**
Commit Message:

contrib: add QAT private key provider for RSA handshake acceleration.

Next generation Intel® QAT support with Intel® Xeon® Scalable processors will feature an Intel® QAT cryptography and compression acceleration engine.

QAT private key provider extension will use qatlib library (https://github.com/intel/qatlib) to accelerate RSA operations in handshakes. The extension will look a bit like the existing cryptomb private key provider. The use case is to move the expensive cryptographic operations away from the CPU to the accelerator device, leaving CPU cycles for other use.

Additional Description:

Support for Intel® QAT is already present in the mainline Linux kernel and in Kubernetes device plugins (to expose the device files to containers). There are previous generations of Intel QAT® hardware devices, but they are not supported by this extension.

Two patches to libqat are present, but they should be included into the next qatlib release.

Risk Level: medium (contrib extension, no HW available yet, crypto-related code)
Testing: unit tests
Docs Changes: Protobuf comments
Release Notes: N/A
Platform Specific Features: Requires a future Intel Xeon Scalable processor.
Fixes #21531 (partly, QATzip extension will be a later PR)

### CODE REVIEW

File: api/contrib/envoy/extensions/private_key_providers/qat/v3alpha/qat.proto [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: `QAT` private key provider]
+// [#extension: envoy.tls.key_providers.qat]
+
+// This message specifies how the private key provider is configured.
+// The private key provider provides RSA sign and decrypt operation
+// hardware acceleration.
+
+// [#extension-category: envoy.tls.key_providers]
```
Comment (daixiang0) [inline api/contrib/envoy/extensions/private_key_providers/qat/v3alpha/qat.proto:26]: Please remove this line as we do need one `envoy.tls.key_providers` category.

Comment (ipuustin) [reply api/contrib/envoy/extensions/private_key_providers/qat/v3alpha/qat.proto]: Ok, fixed.

Comment (wrowe) [reply api/contrib/envoy/extensions/private_key_providers/qat/v3alpha/qat.proto]: Ping @daixiang0 if you are able to re-review and stamp

**Comment by wrowe** [comment]
/assign-from @envoyproxy/senior-maintainers 

for new contrib component review

**Comment by ggreenway** [comment]
@moderation how does this look for dependencies?

**Comment by moderation** [comment]
Not great but this is contrib and follows prior Intel contributions. Lack of code review, contributors and maintenance noted.

scorecard --repo=https://github.com/intel/qatlib

```
RESULTS
-------
Aggregate score: 5.4 / 10

Check scores:
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
|  SCORE  |          NAME          |             REASON             |                             DOCUMENTATION/REMEDIATION                             |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts       | no binaries found in the repo  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#binary-artifacts       |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 6 / 10  | Branch-Protection      | branch protection is not       | https://github.com/ossf/scorecard/blob/main/docs/checks.md#branch-protection      |
|         |                        | maximal on development and all |                                                                                   |
|         |                        | release branches               |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | CI-Tests               | 0 out of 2 merged PRs          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#ci-tests               |
|         |                        | checked by a CI test -- score  |                                                                                   |
|         |                        | normalized to 0                |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | CII-Best-Practices     | no badge detected              | https://github.com/ossf/scorecard/blob/main/docs/checks.md#cii-best-practices     |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 1 / 10  | Code-Review            | GitHub code reviews found for  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#code-review            |
|         |                        | 2 commits out of the last 17   |                                                                                   |
|         |                        | -- score normalized to 1       |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Contributors           | 0 different organizations      | https://github.com/ossf/scorecard/blob/main/docs/checks.md#contributors           |
|         |                        | found -- score normalized to 0 |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Dangerous-Workflow     | no dangerous workflow patterns | https://github.com/ossf/scorecard/blob/main/docs/checks.md#dangerous-workflow     |
|         |                        | detected                       |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Dependency-Update-Tool | no update tool detected        | https://github.com/ossf/scorecard/blob/main/docs/checks.md#dependency-update-tool |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Fuzzing                | project is not fuzzed          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#fuzzing                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | License                | license file detected          | https://github.com/ossf/scorecard/blob/main/docs/checks.md#license                |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | Maintained             | 0 commit(s) out of 17 and 0    | https://github.com/ossf/scorecard/blob/main/docs/checks.md#maintained             |
|         |                        | issue activity out of 5 found  |                                                                                   |
|         |                        | in the last 90 days -- score   |                                                                                   |
|         |                        | normalized to 0                |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| ?       | Packaging              | no published package detected  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#packaging              |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Pinned-Dependencies    | all dependencies are pinned    | https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies    |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 0 / 10  | SAST                   | SAST tool is not run on all    | https://github.com/ossf/scorecard/blob/main/docs/checks.md#sast                   |
|         |                        | commits -- score normalized to |                                                                                   |
|         |                        | 0                              |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Security-Policy        | security policy file detected  | https://github.com/ossf/scorecard/blob/main/docs/checks.md#security-policy        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| ?       | Signed-Releases        | no releases found              | https://github.com/ossf/scorecard/blob/main/docs/checks.md#signed-releases        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Token-Permissions      | tokens are read-only in GitHub | https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions      |
|         |                        | workflows                      |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| 10 / 10 | Vulnerabilities        | no vulnerabilities detected    | https://github.com/ossf/scorecard/blob/main/docs/checks.md#vulnerabilities        |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
| ?       | Webhooks               | check is not supported for     | https://github.com/ossf/scorecard/blob/main/docs/checks.md#webhooks               |
|         |                        | this request: SCORECARD_V6     |                                                                                   |
|         |                        | is not set, not running the    |                                                                                   |
|         |                        | Webhook check                  |                                                                                   |
|---------|------------------------|--------------------------------|-----------------------------------------------------------------------------------|
```

**Comment by phlax** [comment]
@ipuustin this seems to have introduced a flake in CI:

https://dev.azure.com/cncf/envoy/_build/results?buildId=113225&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=6b0ace90-28b2-51d9-2951-b1fd35e67dec&l=2212

**Comment by ipuustin** [comment]
> @ipuustin this seems to have introduced a flake in CI:
> 
> https://dev.azure.com/cncf/envoy/_build/results?buildId=113225&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=6b0ace90-28b2-51d9-2951-b1fd35e67dec&l=2212

Thanks for the heads up. I'll take a look at `ops_test` to see what might go wrong there.

**Pull request merged**